### PR TITLE
Lightweight journal

### PR DESCRIPTION
**Pull request by scottschurr**
This is a fairly mechanical pull request, but it touches a lot of code.  Users of `beast::Journal` must now call member functions, instead of invoking public data members, to access `Streams` at various `Severity` thresholds.  At the beginning of this exercise each `Journal` object carries a reference and 6 small classes as data members.  At the end of the exercise a `Journal` object carries a single reference as its only data member.  So `Journal` becomes a much better choice for pass-by-value.

To make the new member functions as efficient as the old code, at least in release builds, the `Journal::ScopedStream`, `Journal::Stream`, and `Journal` classes were re-worked a bit with close attention to what should be in-lined and what should not.  I looked at the clang optimized disassembly to guide me in these decisions. It was important that the `active()` check in a `JLOG` could be completed with a single call to a virtual function.  No additional constructors or destructors are called, at least in optimized code.

Some of the preferred names for the member functions collided with the names of the public data members, so the rename had to be done in several steps:

 1. Introduce the `Journal` member functions using less desirable names.
 2. Switch all users of the `Journal` data members to call the new member functions.
 3. Remove the data members and provide `Journal` member functions with better names.
 4. Switch all callers of the less desirable names to call the better names.
 5. Remove the member functions with less desirable names from `Journal`.

Much of the renaming was done with python scripts.  But some of the renaming was manual.  The manual renaming allowed me to check for cases where `JLOG` (or some equivalent construct) was not being used.  Those cases are fixed.

I'd like for the reviewers to think about how we should push the end result into develop.  For the pull request I kept the automated changes separate from the manual changes.  For the final commit I'd be happy to `[FOLD]` any or all of the commits.  Let me know what you think.

I'm not assigning reviewers because I'm hoping for volunteers.  A lot has changed, but much of it is mechanical.

### CODE REVIEW

File: src/beast/beast/utility/Journal.h [hunk 1]
```diff
@@ -54,11 +43,30 @@ class Journal
...
+
+    It is advisable to check Journal::active(level) prior to formatting log
+    text.  Doing so sidesteps expensive text formatting when the results
+    will not be sent to the log.
+*/
+class Journal
+{
+public:
+     /** Severity level / threshold of a Journal message. */
+    using Severity = severities::Severity;
```
Comment (seelabs) [inline src/beast/beast/utility/Journal.h]: Nit: Make this `using` private?

Comment (scottschurr) [inline src/beast/beast/utility/Journal.h]: Good idea.  I think I've already switched all of the users of `Severity` to reference `severities::Severity`

File: src/beast/beast/utility/Journal.h [hunk 2]
```diff
@@ -70,6 +78,7 @@ class Journal
         Sink () = delete;
         explicit Sink(Sink const& sink) = default;
         Sink (Severity thresh, bool console);
+        Sink& operator= (Sink& lhs) = delete;
```
Comment (seelabs) [inline src/beast/beast/utility/Journal.h]: Need a `const` on the param: I.e. `Sink& operator=(Sink const&) = delete` (what's here will still give an error if someone tried assign a `const sink` to a sink, but not the one we want).

File: src/ripple/app/ledger/Ledger.cpp [hunk 3]
```diff
@@ -1104,7 +1109,8 @@ bool pendSaveValidated (
 
     if (! app.pendingSaves().shouldWork (ledger->info().seq, isSynchronous))
     {
-        JLOG (app.journal ("Ledger").debug)
+        auto stream = app.journal ("Ledger").debug();
+        JLOG (stream)
```
Comment (seelabs) [inline src/ripple/app/ledger/Ledger.cpp]: I tried to modify `JLOG` so we could write `JLOG(app.journal("Ledger").debug())` and only evaluate the expression once. I ran into trouble assigning to a variable in the conditional and inverting the conditional. It's OK as is.

File: src/beast/beast/utility/Journal.h (hunk 2 above)
Comment (scottschurr) [inline src/beast/beast/utility/Journal.h]: Absolutely!  Good catch.  Thanks.

File: src/beast/beast/utility/Journal.h [hunk 4]
```diff
@@ -107,76 +123,98 @@ class Journal
 
     class Stream;
 
-    /** Scoped ostream-based container for writing messages to a Journal. */
+private:
+    /* Scoped ostream-based container for writing messages to a Journal. */
```
Comment (seelabs) [inline src/beast/beast/utility/Journal.h]: Not in your change set. I'll note that `Journal` has three nested classes: `ScopedStream`, `Stream`, and  `Sink`. It's easy to get lost in the definition of `Journal` with three largish nested classes. I'd rather have them in a detail namespace or somesuch (no change needed).

**Comment by seelabs** [comment]
:+1: Nice structuring of the PR into commit. As for merging into develop, I'd squash all the commits down to one. This change is one logical commit.

**Comment by scottschurr** [comment]
Addressed review comments.  Thanks catching those, @seelabs.

**Comment by miguelportilla** [comment]
The Stream::active function can be removed if its bool operator returns m_sink.active(m_level);

**Comment by miguelportilla** [comment]
:+1:  @scottschurr Nice job. It would be nice if we could avoid creating Stream objects to figure if they are active but I can't find an elegant and simple way of doing that without changing everything.

**Comment by scottschurr** [comment]
@miguelportilla Thanks.  Yeah, it bugs me a bit that we construct a `Stream` at every JLOG occurrence (twice if we actually write to the log).  With the current design the best option is to make the `Stream` construction as light weight as possible.  So that's what I've attempted.  `Journal` and `Stream` are quite light weight, both in data and code overhead.  Constructing a `ScopedStream` is postponed until we know we're doing string formatting.  Once we're string formatting there's no point in trying to be efficient :wink:.

I left `Stream::active()` in the code base even though I think nothing but `Stream::operator bool()` calls it.  It's kind of nice to have a name for the question that `Stream::operator bool()` answers, but I don't feel strongly about it.  What do you think?  Should I remove `Stream::active()`?

Thanks for the review.

**Comment by vinniefalco** [comment]
@scottschurr I would leave `active` in there so caching the value in a local variable reads naturally.

**Comment by scottschurr** [comment]
@vinniefalco Thanks.  That's my inclination as well.

**Comment by scottschurr** [comment]
Squashed.

**Comment by miguelportilla** [comment]
@scottschurr I am okay either way with active. Thanks for sharing your thoughts on the design decisions.

**Pull request closed without merging**